website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,764 @@
+require('css/registrations.css');
+
+var $ = require('jquery');
+var ko = require('knockout');
+var bootbox = require('bootbox');
+var moment = require('moment');
+var URI = require('URIjs');
+
+require('js/koHelpers');
+
+var $osf = require('js/osfHelpers');
+var language = require('js/osfLanguage').registrations;
+
+var editorExtensions = require('js/registrationEditorExtensions');
+
+/**
+ * @class Question
+ * Model for schema questions
+ *
+ * @param {Object} data: optional instantiation values
+ * @param {String} data.title
+ * @param {String} data.nav: short version of title
+ * @param {String} data.type: 'string' | 'number' | 'choose' | 'object'; data type
+ * @param {String} data.format: 'text' | 'textarea' | 'list'; format corresponding with data.type
+ * @param {String} data.description
+ * @param {String} data.help
+ * @param {String} data.required
+ * @param {String[]} data.options: array of options for 'choose' types
+ * @param {Object[]} data.properties: object of sub-Question properties for 'object' types
+ * @param {String} data.match: optional string that must be matched
+ * @param {String} id: unique identifier
+ **/
+var Question = function(data, id) {
+    var self = this;
+
+    self.id = id || -1;
+
+    if (data.value && typeof(data.value) === 'function') {
+        // For subquestions, this could be an observable
+        _value = data.value();
+    } else {
+        _value = data.value || null;
+    }
+    self.value = ko.observable(_value);
+    self.setValue = function(val) {
+        self.value(val);
+    };
+
+    self.title = data.title || 'Untitled';
+    self.nav = data.nav || 'Untitled';
+    self.type = data.type || 'string';
+    self.format = data.format || 'text';
+    self.required = data.required || false;
+    self.description = data.description || '';
+    self.help = data.help;
+    self.options = data.options || [];
+    self.properties = data.properties || {};
+    self.match = data.match || '';
+
+    if (self.required) {
+        self.value.extend({
+            required: true
+        });
+    } else {
+        self.value.extend({
+            required: false
+        });
+    }
+    // A computed to allow rate-limiting save calls
+    self.delayedValue = ko.computed(function() {
+        return self.value();
+    }).extend({ rateLimit: { method: 'notifyWhenChangesStop', timeout: 1000 } });
```

Comment:
Although this limits how often requests are sent, it doesn't guarantee that only one request is pending at a time. 

We'd like to avoid a situation where multiple pending requests might get processed out of order (eg the middle of a sentence being saved after the end of a sentence, causing data loss). Will discuss further with @lyndsysimon when he's completed the pending changes discussed earlier.